apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,43 +63,115 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
         /// <param name="consumer">The consumer to flush to, cannot be null.</param>
         /// <param name="bufferSize">The buffer size in bytes.</param>
         public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T> sizer, int bufferSize)
+            => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
         {
-            return new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _semaphore.Wait();
+                foreach (T item in items)
+                {
+                    _size += _sizer.GetSize(item);
+                    _items.Add(item);
+                    if (_size >= _bufferSize)
+                    {
+                        FlushesWithSemaphoreHeld();
+                    }
+                }
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
```

Comment:
Actually, why not just go back to use locks and use Monitor.Enter/Exit using the pattern you describe and shown [here](https://msdn.microsoft.com/en-us/library/dd289498(v=vs.110).aspx)? I see the comment about releasing on different threads, but none of the await calls use `ConfigureAwait(false)` (should they be?), so won't those calls be on the same thread?